appserver/extras/embedded/shell/glassfish-embedded-static-shell/pom.xml

```diff
@@ -671,6 +671,12 @@
             <groupId>javax.enterprise</groupId>
             <artifactId>cdi-api</artifactId>
             <optional>true</optional>
+            <exclusions>
```

Comment:
put this under appserver/pom.xml in the dependency management section.

Reply:
This cannot be there. That would prevent the code from being built. Here it is to have proper classpath in Manifest.MF. Code in this module does not need javax.inject, so it compiles just fine. But it adds the right classpath to jakarta.inject.jar only when the javax.inject is not found by maven. 
This is no hiding any real issue. There is no issue. Just fixing classpath.

Reply:
Again, should you add the exclusion to the dependency management, it will exclude the javax.inject in every cdi dependency. That is NOT what you want. You need the dependency. You do not want the dependency just in these two places.

Reply:
Does not make sense to add it only to 2 places. Fix it up once for every usage of cdi-api and add a dependency on jakarta.inject where needed.

Note that glassfish-embedded-static-shell is actually one of the use-cases where class-path matters. You previously said there was no actual issue other than class-path, now you're saying there is something wrong without providing details ?!

Reply:
Quite better :)

./appserver/deployment/dol/pom.xml has the version of cdi-api hard-coded to `2.0`, it should be removed since the dependency is managed.

Now is the time to add missing direct dependencies to jakarta-inject, we can do this for all modules not only the ones that depend on cdi-api. I did some grepping and the list of all modules using javax.inject in their code is below, none of them seem to have a direct dependency on javax.inject or jakarta.inject.

```
./appserver/admin/admin-core/pom.xml
./appserver/admingui/plugin-service/pom.xml
./appserver/appclient/client/acc/pom.xml
./appserver/appclient/server/connector/pom.xml
./appserver/appclient/server/core/pom.xml
./appserver/batch/glassfish-batch-commands/pom.xml
./appserver/batch/glassfish-batch-connector/pom.xml
./appserver/common/amx-javaee/pom.xml
./appserver/common/annotation-framework/pom.xml
./appserver/common/container-common/pom.xml
./appserver/common/glassfish-naming/pom.xml
./appserver/concurrent/concurrent-impl/pom.xml
./appserver/connectors/admin/pom.xml
./appserver/connectors/connectors-connector/pom.xml
./appserver/connectors/connectors-inbound-runtime/pom.xml
./appserver/connectors/connectors-internal-api/pom.xml
./appserver/connectors/connectors-runtime/pom.xml
./appserver/connectors/work-management/pom.xml
./appserver/core/javaee-kernel/pom.xml
./appserver/deployment/dol/pom.xml
./appserver/deployment/javaee-core/pom.xml
./appserver/deployment/javaee-full/pom.xml
./appserver/ejb/ejb-connector/pom.xml
./appserver/ejb/ejb-container/pom.xml
./appserver/ejb/ejb-full-container/pom.xml
./appserver/grizzly/grizzly-container/pom.xml
./appserver/ha/ha-shoal-cache-bootstrap/pom.xml
./appserver/jdbc/admin/pom.xml
./appserver/jdbc/jdbc-runtime/pom.xml
./appserver/jms/admin/pom.xml
./appserver/jms/gf-jms-injection/pom.xml
./appserver/jms/jms-core/pom.xml
./appserver/load-balancer/admin/pom.xml
./appserver/load-balancer/gf-load-balancer-connector/pom.xml
./appserver/orb/orb-connector/pom.xml
./appserver/orb/orb-enabler/pom.xml
./appserver/orb/orb-iiop/pom.xml
./appserver/persistence/cmp/support-ejb/pom.xml
./appserver/persistence/cmp/support-sqlstore/pom.xml
./appserver/persistence/entitybean-container/pom.xml
./appserver/persistence/gf-jpa-connector/pom.xml
./appserver/persistence/jpa-container/pom.xml
./appserver/resources/javamail/javamail-connector/pom.xml
./appserver/resources/javamail/javamail-runtime/pom.xml
./appserver/resources/resources-connector/pom.xml
./appserver/resources/resources-runtime/pom.xml
./appserver/security/appclient.security/pom.xml
./appserver/security/core-ee/pom.xml
./appserver/security/ejb.security/pom.xml
./appserver/security/webintegration/pom.xml
./appserver/security/webservices.security/pom.xml
./appserver/transaction/internal-api/pom.xml
./appserver/transaction/jta/pom.xml
./appserver/transaction/jts/pom.xml
./appserver/verifier/verifier-impl/pom.xml
./appserver/web/admin/pom.xml
./appserver/web/gf-web-connector/pom.xml
./appserver/web/gf-weld-connector/pom.xml
./appserver/web/gui-plugin-common/pom.xml
./appserver/web/jersey-mvc-connector/pom.xml
./appserver/web/jsf-connector/pom.xml
./appserver/web/jspcaching-connector/pom.xml
./appserver/web/jstl-connector/pom.xml
./appserver/web/war-util/pom.xml
./appserver/web/web-core/pom.xml
./appserver/web/web-glue/pom.xml
./appserver/web/web-ha/pom.xml
./appserver/web/web-sse/pom.xml
./appserver/web/weld-integration/pom.xml
./appserver/webservices/connector/pom.xml
./appserver/webservices/jsr109-impl/pom.xml
./appserver/webservices/metro-glue/pom.xml
./nucleus/admin/cli/pom.xml
./nucleus/admin/config-api/pom.xml
./nucleus/admin/monitor/pom.xml
./nucleus/admin/rest/gf-restadmin-connector/pom.xml
./nucleus/admin/rest/rest-service/pom.xml
./nucleus/admin/server-mgmt/pom.xml
./nucleus/admin/util/pom.xml
./nucleus/cluster/admin/pom.xml
./nucleus/cluster/cli/pom.xml
./nucleus/cluster/gms-adapter/pom.xml
./nucleus/cluster/gms-bootstrap/pom.xml
./nucleus/common/amx-core/pom.xml
./nucleus/common/common-util/pom.xml
./nucleus/common/glassfish-api/pom.xml
./nucleus/common/internal-api/pom.xml
./nucleus/common/mbeanserver/pom.xml
./nucleus/core/bootstrap/pom.xml
./nucleus/core/kernel/pom.xml
./nucleus/core/logging/pom.xml
./nucleus/deployment/admin/pom.xml
./nucleus/deployment/autodeploy/pom.xml
./nucleus/deployment/common/pom.xml
./nucleus/diagnostics/context/pom.xml
./nucleus/flashlight/framework/pom.xml
./nucleus/grizzly/config/pom.xml
./nucleus/hk2/config-types/pom.xml
./nucleus/hk2/hk2-config/pom.xml
./nucleus/osgi-platforms/osgi-cli-interactive/pom.xml
./nucleus/osgi-platforms/osgi-cli-remote/pom.xml
./nucleus/osgi-platforms/osgi-container/pom.xml
./nucleus/resources/pom.xml
./nucleus/security/core/pom.xml
./nucleus/security/services/pom.xml
./nucleus/security/ssl-impl/pom.xml
./nucleus/tests/admin/pom.xml
```

It may be too much to ask at this point, let me know.  We can file a bug and do it separately with another PR.

Reply:
It's a Maven best practice: define compile time dependencies as direct dependencies. 

Otherwise it is sensitive to the transitive dependencies. In the snippet you shared it's nested 3 levels down. If another dependency has a bad version of javax.inject 2 levels down, the build class-path would have the bad one before the good one ;  same thing with the maven-bundle-plugin.

Reply:
So should I file the bug, or will you do it?

Reply:
I'll do it. I don't want to impose that on you ;)

Reply:
filed https://github.com/eclipse-ee4j/glassfish/issues/22682